# Практическая работа

# Задача

Один из способов повысить эффективность взаимодействия банка с клиентами — отправлять предложение о новой услуге не всем клиентам, а только некоторым, которые выбираются по принципу наибольшей склонности к отклику на это предложение.

Задача заключается в том, чтобы предложить алгоритм, который будет выдавать склонность клиента к положительному или отрицательному отклику на предложение банка. Предполагается, что, получив такие оценки для некоторого множества клиентов, банк обратится с предложением только к тем, от кого ожидается положительный отклик.


Для решения этой задачи загрузите файлы из базы в Postgres.
Эта БД хранит информацию о клиентах банка и их персональные данные, такие как пол, количество детей и другие.

Описание таблиц с данными представлено ниже.


**D_work**

Описание статусов относительно работы:
- ID — идентификатор социального статуса клиента относительно работы;
- COMMENT — расшифровка статуса.


**D_pens**

Описание статусов относительно пенсии:
- ID — идентификатор социального статуса;
- COMMENT — расшифровка статуса.


**D_clients**

Описание данных клиентов:
- ID — идентификатор записи;
- AGE	— возраст клиента;
- GENDER — пол клиента (1 — мужчина, 0 — женщина);
- EDUCATION — образование;
- MARITAL_STATUS — семейное положение;
- CHILD_TOTAL	— количество детей клиента;
- DEPENDANTS — количество иждивенцев клиента;
- SOCSTATUS_WORK_FL	— социальный статус клиента относительно работы (1 — работает, 0 — не работает);
- SOCSTATUS_PENS_FL	— социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
- REG_ADDRESS_PROVINCE — область регистрации клиента;
- FACT_ADDRESS_PROVINCE — область фактического пребывания клиента;
- POSTAL_ADDRESS_PROVINCE — почтовый адрес области;
- FL_PRESENCE_FL — наличие в собственности квартиры (1 — есть, 0 — нет);
- OWN_AUTO — количество автомобилей в собственности.


**D_agreement**

Таблица с зафиксированными откликами клиентов на предложения банка:
- AGREEMENT_RK — уникальный идентификатор объекта в выборке;
- ID_CLIENT — идентификатор клиента;
- TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было).
    
    
**D_job**

Описание информации о работе клиентов:
- GEN_INDUSTRY — отрасль работы клиента;
- GEN_TITLE — должность;
- JOB_DIR — направление деятельности внутри компании;
- WORK_TIME — время работы на текущем месте (в месяцах);
- ID_CLIENT — идентификатор клиента.


**D_salary**

Описание информации о заработной плате клиентов:
- ID_CLIENT — идентификатор клиента;
- FAMILY_INCOME — семейный доход (несколько категорий);
- PERSONAL_INCOME — личный доход клиента (в рублях).


**D_last_credit**

Информация о последнем займе клиента:
- ID_CLIENT — идентификатор клиента;
- CREDIT — сумма последнего кредита клиента (в рублях);
- TERM — срок кредита;
- FST_PAYMENT — первоначальный взнос (в рублях).


**D_loan**

Информация о кредитной истории клиента:
- ID_CLIENT — идентификатор клиента;
- ID_LOAN — идентификатор кредита.

**D_close_loan**

Информация о статусах кредита (ссуд):
- ID_LOAN — идентификатор кредита;
- CLOSED_FL — текущий статус кредита (1 — закрыт, 0 — не закрыт).

Ниже представлен минимальный список колонок, которые должны находиться в итоговом датасете после склейки и агрегации данных. По своему усмотрению вы можете добавить дополнительные к этим колонки.

    - AGREEMENT_RK — уникальный идентификатор объекта в выборке;
    - TARGET — целевая переменная: отклик на маркетинговую кампанию (1 — отклик был зарегистрирован, 0 — отклика не было);
    - AGE — возраст клиента;
    - SOCSTATUS_WORK_FL — социальный статус клиента относительно работы (1 — работает, 0 — не работает);
    - SOCSTATUS_PENS_FL — социальный статус клиента относительно пенсии (1 — пенсионер, 0 — не пенсионер);
    - GENDER — пол клиента (1 — мужчина, 0 — женщина);
    - CHILD_TOTAL — количество детей клиента;
    - DEPENDANTS — количество иждивенцев клиента;
    - PERSONAL_INCOME — личный доход клиента (в рублях);
    - LOAN_NUM_TOTAL — количество ссуд клиента;
    - LOAN_NUM_CLOSED — количество погашенных ссуд клиента.


Будьте внимательны при сборке датасета: это реальные банковские данные, в которых могут наблюдаться дубли, некорректно заполненные значения или значения, противоречащие друг другу. Для получения качественной модели необходимо предварительно очистить датасет от такой информации.

## Задание 1

Соберите всю информацию о клиентах в одну таблицу, где одна строчка соответствует полной информации об одном клиенте.

## Задание 2

При помощи инструмента Streamlit проведите разведочный анализ данных. В него может входить:

* построение графиков распределений признаков
* построение матрицы корреляций
* построение графиков зависимостей целевой переменной и признаков
* вычисление числовых характеристик распределения числовых столбцов (среднее, min, max, медиана и так далее)
* любые другие ваши идеи приветствуются!

[Пример Streamlit-приложения](https://rateyourflight.streamlit.app) с разведочным анализом, прогнозом модели и оценкой ее результатов.

In [1]:
import pandas as pd

In [2]:
#Загружаем данные
df1 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_clients.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_close_loan.csv')
df3 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_job.csv')
df4 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_last_credit.csv')
df5 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_loan.csv')
df6 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_pens.csv')
df7 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_salary.csv')
df8 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_target.csv')
df9 = pd.read_csv('https://raw.githubusercontent.com/aiedu-courses/stepik_linear_models/main/datasets/D_work.csv')

In [8]:


#Объединяем данные
merged_df = df3.merge(df4, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df5, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df7, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df8, on='ID_CLIENT', how='outer')
merged_df = merged_df.merge(df2, on='ID_LOAN', how='outer')
df1.rename(columns={'ID': 'ID_CLIENT'}, inplace=True)
merged_df = merged_df.merge(df1, on='ID_CLIENT', how='outer')

#Удаляем полные дубликаты и делаем бинарные переменные
merged_df = merged_df.drop_duplicates()

# Создание столбца LOAN_NUM_TOTAL
merged_df['LOAN_NUM_TOTAL'] = merged_df.groupby('ID_CLIENT')['ID_LOAN'].transform('count')
# Создание столбца LOAN_NUM_CLOSED
merged_df['LOAN_NUM_CLOSED'] = merged_df.groupby('ID_CLIENT')['CLOSED_FL'].transform(lambda x: (x == 1).sum())

#Удаление разных кредитов одного клиента
merged_df.drop_duplicates(subset=['ID_CLIENT'], keep='first', inplace=True)

#Отбираем нужные столбцы
columns = ['AGREEMENT_RK','TARGET','AGE','SOCSTATUS_WORK_FL','SOCSTATUS_PENS_FL','GENDER','CHILD_TOTAL','DEPENDANTS','PERSONAL_INCOME','LOAN_NUM_TOTAL','LOAN_NUM_CLOSED']
merged_df = merged_df[columns]



In [9]:
# Проверки нуловых значений
merged_df.isna().sum()

AGREEMENT_RK         777
TARGET               777
AGE                    0
SOCSTATUS_WORK_FL      0
SOCSTATUS_PENS_FL      0
GENDER                 0
CHILD_TOTAL            0
DEPENDANTS             0
PERSONAL_INCOME      777
LOAN_NUM_TOTAL         0
LOAN_NUM_CLOSED        0
dtype: int64

In [11]:
merged_df[merged_df.AGREEMENT_RK.isna()].describe()

AGREEMENT_RK  TARGET         AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
count           0.0     0.0  777.000000         777.000000         777.000000   
mean            NaN     NaN   40.172458           0.904762           0.137709   
std             NaN     NaN   11.591795           0.293733           0.344816   
min             NaN     NaN   21.000000           0.000000           0.000000   
25%             NaN     NaN   30.000000           1.000000           0.000000   
50%             NaN     NaN   39.000000           1.000000           0.000000   
75%             NaN     NaN   50.000000           1.000000           0.000000   
max             NaN     NaN   65.000000           1.000000           1.000000   

           GENDER  CHILD_TOTAL  DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  \
count  777.000000   777.000000  777.000000              0.0           777.0   
mean     0.628057     1.061776    0.629344              NaN             0.0   
std      0.483635     1.001953    0.806906              NaN             0.0   
min      0.000000     0.000000    0.000000              NaN             0.0   
25%      0.000000     0.000000    0.000000              NaN             0.0   
50%      1.000000     1.000000    0.000000              NaN             0.0   
75%      1.000000     2.000000    1.000000              NaN             0.0   
max      1.000000     7.000000    7.000000              NaN             0.0   

       LOAN_NUM_CLOSED  
count            777.0  
mean               0.0  
std                0.0  
min                0.0  
25%                0.0  
50%                0.0  
75%                0.0  
max                0.0

У 777 объектов нет таргета, а также несколих других колонок, удалим такие объекты.

In [12]:
merged_df = merged_df.dropna()
merged_df.reset_index(drop=True, inplace=True)

In [13]:
merged_df.describe()

AGREEMENT_RK        TARGET           AGE  SOCSTATUS_WORK_FL  \
count  1.522300e+04  15223.000000  15223.000000       15223.000000   
mean   6.540183e+07      0.119030     40.406096           0.909610   
std    4.568181e+06      0.323835     11.601068           0.286748   
min    5.991015e+07      0.000000     21.000000           0.000000   
25%    6.192092e+07      0.000000     30.000000           1.000000   
50%    6.437100e+07      0.000000     39.000000           1.000000   
75%    6.708802e+07      0.000000     50.000000           1.000000   
max    7.529224e+07      1.000000     67.000000           1.000000   

       SOCSTATUS_PENS_FL        GENDER   CHILD_TOTAL    DEPENDANTS  \
count       15223.000000  15223.000000  15223.000000  15223.000000   
mean            0.134468      0.654536      1.099389      0.645208   
std             0.341165      0.475535      0.995411      0.812252   
min             0.000000      0.000000      0.000000      0.000000   
25%             0.000000      0.000000      0.000000      0.000000   
50%             0.000000      1.000000      1.000000      0.000000   
75%             0.000000      1.000000      2.000000      1.000000   
max             1.000000      1.000000     10.000000      7.000000   

       PERSONAL_INCOME  LOAN_NUM_TOTAL  LOAN_NUM_CLOSED  
count     15223.000000    15223.000000     15223.000000  
mean      13853.836323        1.387769         0.751889  
std        9015.467617        0.793566         0.988499  
min          24.000000        1.000000         0.000000  
25%        8000.000000        1.000000         0.000000  
50%       12000.000000        1.000000         0.000000  
75%       17000.000000        2.000000         1.000000  
max      250000.000000       11.000000        11.000000

Есть подозрительные моменты, например:

* 10 детей
* 7 иждивенцев
* доход - 24

Проверим такие объекты.

In [16]:
ind = merged_df.CHILD_TOTAL > 5
merged_df[ind]

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  \
2385     61250208.0     0.0   49                  1                  0   
2916     61382118.0     0.0   60                  0                  1   
4017     61958911.0     1.0   56                  0                  1   
4976     62249518.0     0.0   60                  1                  1   
5071     62261659.0     0.0   62                  0                  1   
5183     62377743.0     0.0   59                  0                  1   
6672     63584425.0     0.0   56                  1                  1   
6968     63894742.0     0.0   39                  1                  0   
7462     64343517.0     0.0   36                  1                  0   
7717     64428091.0     0.0   56                  0                  1   
7736     64429995.0     0.0   39                  1                  0   
8069     64559789.0     0.0   59                  1                  0   
8614     64913008.0     1.0   39                  1                  0   
9948     66589572.0     0.0   64                  0                  1   
12251    67335240.0     0.0   43                  1                  0   
12432    67431060.0     0.0   59                  1                  0   
13115    73833445.0     0.0   50                  1                  0   
13216    73845860.0     0.0   61                  0                  1   
13377    74025823.0     0.0   53                  0                  1   
13936    74523153.0     0.0   58                  1                  1   
14171    74641291.0     0.0   56                  1                  1   

       GENDER  CHILD_TOTAL  DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  \
2385        1            6           0          12000.0               1   
2916        1            6           0           5200.0               2   
4017        1            6           0           5000.0               2   
4976        1            6           0           7000.0               1   
5071        0            6           0           7000.0               1   
5183        1            7           0          12000.0               1   
6672        1            8           1           7500.0               1   
6968        1            6           6          12000.0               1   
7462        0            6           6          20000.0               1   
7717        1            7           1           4500.0               1   
7736        0            6           6          15000.0               1   
8069        0            7           0          10000.0               1   
8614        1           10           0           8500.0               1   
9948        1           10           1           4200.0               3   
12251       1            6           3          12000.0               1   
12432       0            6           1          10000.0               1   
13115       0            7           7          24000.0               1   
13216       1           10           0           7300.0               1   
13377       1            7           1           4400.0               1   
13936       1            6           0           9000.0               1   
14171       1            6           1          18700.0               6   

       LOAN_NUM_CLOSED  
2385                 1  
2916                 1  
4017                 0  
4976                 0  
5071                 1  
5183                 1  
6672                 0  
6968                 0  
7462                 1  
7717                 0  
7736                 0  
8069                 0  
8614                 0  
9948                 3  
12251                0  
12432                0  
13115                0  
13216                1  
13377                0  
13936                1  
14171                5

In [19]:
ind = (merged_df.CHILD_TOTAL > 5) & (merged_df.AGE - (18 + merged_df.CHILD_TOTAL - merged_df.DEPENDANTS) < 20)
merged_df[ind]

AGREEMENT_RK  TARGET  AGE  SOCSTATUS_WORK_FL  SOCSTATUS_PENS_FL  GENDER  \
7462    64343517.0     0.0   36                  1                  0       0   
8614    64913008.0     1.0   39                  1                  0       1   

      CHILD_TOTAL  DEPENDANTS  PERSONAL_INCOME  LOAN_NUM_TOTAL  \
7462            6           6          20000.0               1   
8614           10           0           8500.0               1   

      LOAN_NUM_CLOSED  
7462                1  
8614                0

Объект 8614 выглядит как выброс, непонятно как у этого объекта может быть 10 детей и 0 иждивенцев в 39 лет. Удалим его из выборки.

In [22]:
merged_df = merged_df[merged_df.index != 8614].reset_index(drop=True)

In [24]:
merged_df.to_csv('data.csv', index=False)